src/sys/stm_tests.ml

```diff
@@ -0,0 +1,275 @@
+      (match res with
+      | Error err ->
+        (match err with
+        | Sys_error s ->
+          (s = (p complete_path) ^ ": Permission denied") ||
+          (s = (p complete_path) ^ ": File exists" && mem_model fs complete_path) ||
+          (s = (p complete_path) ^ ": No such file or directory" && not (mem_model fs path)) ||
+          if Sys.win32 && not (path_is_a_dir fs complete_path)
+          then s = (p complete_path) ^ ": No such file or directory"
+          else s = (p complete_path) ^ ": Not a directory"
+          | _ -> false)
+        | Ok () -> mem_model fs path && path_is_a_dir fs path && not (mem_model fs complete_path))
+    | Rmdir (path, delete_dir_name), Res ((Result (Unit,Exn),_), res) ->
+      let complete_path = (path @ [delete_dir_name]) in
+      (match res with
+      | Error err ->
+        (match err with
+          | Sys_error s ->
+            (s = (p complete_path) ^ ": Directory not empty" && not (readdir_model fs complete_path = Some [])) ||
+            (s = (p complete_path) ^ ": No such file or directory" && not (mem_model fs complete_path)) ||
+            if Sys.win32 && not (path_is_a_dir fs complete_path) (* if not a directory *)
+            then s = (p complete_path) ^ ": Invalid argument"
+            else s = (p complete_path) ^ ": Not a directory"
+          | _ -> false)
+      | Ok () ->
+          mem_model fs complete_path && path_is_a_dir fs complete_path && readdir_model fs complete_path = Some [])
+    | Readdir path, Res ((Result (Array String,Exn),_), res) ->
+      (match res with
+      | Error err ->
+        (match err with
+          | Sys_error s ->
+            (s = (p path) ^ ": Permission denied") ||
+            (s = (p path) ^ ": No such file or directory" && not (mem_model fs path)) ||
+            if Sys.win32 && not (path_is_a_dir fs path) (* if not a directory *)
+            then s = (p path) ^ ": Invalid argument"
+            else s = (p path) ^ ": Not a directory"
+          | _ -> false)
+      | Ok array_of_subdir ->
+        (* Temporary work around for mingW, see https://github.com/ocaml/ocaml/issues/11829 *)
+        if Sys.win32 && not (mem_model fs path)
+        then array_of_subdir = [||]
+        else
+          (mem_model fs path && path_is_a_dir fs path &&
+          (match readdir_model fs path with
+          | None   -> false
+          | Some l ->
+            List.sort String.compare l
+            = List.sort String.compare (Array.to_list array_of_subdir))))
+    | Touch (_path, _new_file_name, _perm), Res ((Unit,_),_) -> true
+    | _,_ -> false
+end
+
+let uname_os () =
+  let ic = Unix.open_process_in "uname -s" in
+  let os = In_channel.input_line ic in
+  In_channel.close ic;
+  os
+
+module Sys_seq = STM_sequential.Make(SConf)
+module Sys_dom = STM_domain.Make(SConf)
+
+;;
+QCheck_base_runner.run_tests_main [
+    Sys_seq.agree_test              ~count:1000 ~name:"STM Sys test sequential";
+    if Sys.unix && uname_os () = Some "Linux"
```

Comment:
I didn’t find in the PR or commit message why the test behave differently under mac and linux, so I was curious what happens.

Reply:
This was purely based on empirical observations - we have not been able to trigger when running these under Linux :man_shrugging: